validphys2/src/validphys/photon/structure_functions.py

```diff
@@ -0,0 +1,113 @@
+from abc import ABC, abstractmethod
+
+import numpy as np
+import pineappl
+from scipy.interpolate import RectBivariateSpline
+
+from . import constants
+
+
+class StructureFunction(ABC):
+    """Abstract class for the DIS structure functions"""
+
+    @abstractmethod
+    def fxq(self, x, Q):
+        r"""
+        Abstract method returning the value of a DIS structure functions.
+
+        Parameters
+        ----------
+        x : float
+            Bjorken x
+        Q : float
+            DIS hard scale
+
+        Returns
+        -------
+        F_{2,L}: float
+            Structure function F2 or FL
+        """
+        pass
+
+
+class InterpStructureFunction(StructureFunction):
+    """
+    Compute an interpolated structure function convoluting an FKtable
+    with a PDF.
+    """
+
+    def __init__(self, path_to_fktable, pdfs):
+        self.fktable = pineappl.fk_table.FkTable.read(path_to_fktable)
+        x = np.unique(self.fktable.bin_left(1))
+        q2 = np.unique(self.fktable.bin_left(0))
```

Comment:
They are not unique since they are done in the following way:
`x=[x1,x1,...,x1,x2,x2,...,x2,x3,...]`
`q=[q1,q2,...qn,q1,...,qn,q1,...]`
The fact that `np.unique` orders them is good (it would be a problem if it didn't) since they are passed to the interpolationg function that requires the lists to be ordered

Reply:
But can't you use instead `self.fktable.x_grid()`? (and, I'm going to guess, `q_grid()`?)

Reply:
Edit: Ah, because these are the kinematic variables of a 2D distribution and not the fktable bins.

Reply:
you've been quicker than me ;-) its the usual $x_{Bjorken},Q^2$ vs $z,\mu^2$ (as we like to say in yadism)